kotest-property/src/commonMain/kotlin/io/kotest/property/Gen.kt

```diff
@@ -0,0 +1,112 @@
+package io.kotest.property
+
+import kotlin.random.Random
+
+/**
+ * A Generator, or [Gen] is responsible for generating data to be used in property testing.
+ * Each generator will generate data for a specific type <T>.
+ *
+ * There are two supported types of generators - [Arbitrary] and [Progression] - defined
+ * as sub interfaces of this interface.
+ *
+ * An arbitrary is used when you need random values across a large space.
+ * A progression is useful when you want exhaustive values from a small space.
+ *
+ * Both types of generators can be mixed and matched in property tests. For example,
+ * you could test a function with 1000 random positive integers (arbitrary) and every
+ * even number from 0 to 200 (progression).
+ */
+interface Gen<T> {
+
+   /**
+    * @return the values provided by this [Gen] as a lazy list.
+    * @see [PropertyInput]
+    */
+   fun generate(random: Random): Sequence<PropertyInput<T>>
+}
+
+/**
+ * Models the type of inputs that can be generated from an [Arbitrary].
+ *
+ * There are two implementations of [PropertyInput]. The first is a wrapper for values without
+ * a shrinker, which is [PropertyInput.Value].
+ *
+ * The second is a wrapper for values associated with shrinks, which is [PropertyInput.ValueAndShrinker].
+ */
+sealed class PropertyInput<out T> {
+   abstract val value: T
+
+   /**
+    * Models a value that cannot be shrunk, or for which shrinking is not supported.
+    *
+    * @param value the property test input value
+    */
+   data class Value<T>(override val value: T) : PropertyInput<T>()
+
+   /**
+    * Models a value that can be shrunk if the value fails to pass a test.
+    *
+    * @param value the property test input value
+    * @param shrinker a function that returns shrunk values of the input value
+    */
+   data class ValueAndShrinker<T>(
+      override val value: T,
+      val shrinker: () -> List<PropertyInput<T>>
+   ) : PropertyInput<T>()
+
+   companion object {
+      operator fun <T> invoke(value: T) = Value(value)
+      operator fun <T> invoke(value: T, shrinker: Shrinker<T>) = ValueAndShrinker(value, { shrinker.shrink(value) })
+   }
+}
+
+fun <T, U> PropertyInput<T>.map(f: (T) -> U): PropertyInput<U> = when (this) {
+   is PropertyInput.Value<T> -> PropertyInput.Value(f(value))
+   is PropertyInput.ValueAndShrinker<T> -> {
+      val shrinks = { this.shrinker().map { it.map(f) } }
+      PropertyInput.ValueAndShrinker(f(value), shrinks)
+   }
+}
+
+fun <T, U> Gen<T>.map(f: (T) -> U): Gen<U> = object : Gen<U> {
```

Comment:
How would it work when I want to use a more specific `map`?

```
Arbitrary.int().map { it + 1 } 
```
Would IntelliJ suggest both options? I'm not s ure if these maps and filters should be extensions for the base `Gen` type, only for the specifics.

What do you think?

Reply:
We don't have a map for arb.